gpio_nrf5: bugfix: Don't require INT_EDGE to enable INT_DOUBLE_EDGE

SebastianBoe opened the pull request:
The driver has been ignoring INT_DOUBLE_EDGE when INT_EDGE is set,
AFAICT from the documentation and other drivers, e..g CC2650, this is
incorrect behaviour.

~~This patch ensures that INT_DOUBLE_EDGE takes effect even though
INT_EDGE has not been set.~~

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
FYI: @IvanSanchez @cvinayak

anangl commented:
The code change is okay. But the commit description is be a bit confusing. I guess it was intended to say "when INT_EDGE _was not_ set".

cvinayak commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -233,10 +233,10 @@ static int gpio_nrf5_config(struct device *dev,
-			if (flags & GPIO_INT_DOUBLE_EDGE) {
-				config |= GPIOTE_CFG_POL_TOGG;
-			} else if (flags & GPIO_INT_ACTIVE_HIGH) {
+		if (flags & GPIO_INT_DOUBLE_EDGE) {
```
The only board using this upstream is:
boards/x86/quark_se_c1000_devboard/board.c

And this board does set GPIO_INT_EDGE alongwith GPIO_INT_DOUBLE_EDGE.

From the include/gpio.h definition, I conclude the intention of the driver model is to classify GPIO_INT_EDGE or GPIO_INT_LEVEL (represented by BIT 5), and GPIO_INT_DOUBLE_EDGE or GPIO_INT_ACTIVE_LOW or GPIO_INT_ACTIVE_HIGH (represented by a 2-bit value, BIT 6 and 2).

SebastianBoe replied:
Okay, so we interpret the documentation differently.

Perhaps we can ask the author ...

SebastianBoe replied:
Hi,

this PR is assuming that DOUBLE_EDGE does not require EDGE and updating gpio_nrf5.c accordingly. If there are users that have assumed otherwise then they will have to be updated as well, but IMHO that should be done independently of this PR.

I could create issues for the users we know of that have interpreted the API differently though.

cvinayak reviewed (changes requested):
Better document the gpio model and/or fix the quark_se_c1000_devboard too (to implicitly set GPIO_INT_EDGE if only GPIO_INT_DOUBLE_EDGE is selected).

SebastianBoe commented:
"The code change is okay. But the commit description is be a bit confusing. I guess it was intended to say "when INT_EDGE was not set" in the first line."
You are right, my mistake. Fixed now.

SebastianBoe commented:
It's not obvious to me what the desired behaviour is, the behaviour is inconsitent among the implementations:

```
DOUBLE_EDGE does not require EDGE:
atmel_sam3
cc2650
dw
qmsi

DOUBLE_EDGE requires EDGE:
cc32xx
mcux_igpio
mcux
fe310
esp32
nrf5
stm32

```

SebastianBoe commented:
Does there exist a code owner for the GPIO API?

anangl commented:
I don't know what was the original intention of the author. But it seems natural that GPIO_INT_DOUBLE_EDGE should imply GPIO_INT_EDGE (you wouldn't combine it with GPIO_INT_LEVEL), so there is no need to specify both. As well as for GPIO_INT_DOUBLE_EDGE there is no need to specify any GPIO_INT_ACTIVE_* flag.

b0661 commented:
Having dived into the GPIO interrupt interface to implement #5799 I think the flags follow the old oszilloscope control concept:

1) select trigger mode (EDGE/ LEVEL)
2) select trigger level (rising, falling, double edge/ high, low)

When looking at that control concept setting  EDGE automatically with DOUBLE_EDGE is a convenience function but nothing that is implied by the interface itself.

SebastianBoe commented:
"When looking at that control concept setting EDGE automatically with DOUBLE_EDGE is a convenience function but nothing that is implied by the interface itself."

I agree. Thank you for sharing.

I have updated the PR to simply have the docs state that DOUBLE_EDGE requires EDGE.

nashif commented:
this change seems to be obsolete? please close

SebastianBoe committed:
gpio: doc: Be explicit about how EDGE and DOUBLE_EDGE work together

Be explicit about how EDGE and DOUBLE_EDGE interact, see PR #6807 for
details.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
Not obsolete. Just needed to be synced with master.

It is now up-to-date again.

SebastianBoe commented:
@nashif : This is ready to be merged.
